lib/cloud-watch-event-listener.js

```diff
@@ -104,15 +104,104 @@ class CloudWatchEventListener extends events.EventEmitter {
     let region = body.region;
     let id = body.detail['instance-id'];
     let state = body.detail.state;
+    // Not a great name, but this variable is the time that the CloudWatch
+    // Event message was generated by the API, not the time at which we
+    // received it
+    let generated = new Date(body.time);
 
+    // We're logging this information to a database table so that we can
+    // see when these events happen
     try {
-      await this.state.logCloudWatchEvent({region, id, state});
+      await this.state.logCloudWatchEvent({region, id, state, received: generated});
     } catch (err) {
-      // We don't want to block things
+      // We don't want to block things, but let's bubble up the error for the
+      // time being
       this.monitor.reportError(err);
     }
 
-    if (state === 'pending' || state === 'running') {
+    let dbResponse;
+
+    let transaction = await this.state.beginTransaction();
+
+    try {
+      let x = await this.state.listInstances({region, id}, transaction);
+      assert(Array.isArray(x));
+      assert(x.length === 0 || x.length === 1);
+      if (x.length === 1) {
+        dbResponse = x[0];
+      }
+    } catch (err) {
+      await this.state.rollbackTransaction(transaction);
+      log.error(err, 'Error looking up state from database');
+    }
+
+    // We want to close this transaction if we didn't find anything in the database.
+    // We'll use a new transaction if we're looking up from the API, but if we got
+    // the information from the database, let's store it with the same transaction
+    if (!dbResponse) {
+      await this.state.commitTransaction(transaction);
+    }
+
+    if (dbResponse) {
+      // If there's a response from the database, we know that we already have
+      // this information in the database.  We'll use this information to
+      // figure out the immutable metadata
+
+      // Simple assertions, so may as well check for them.  If these failed,
+      // something would be foundationally wrong
+
+      assert(dbResponse.id === id);
+      assert(dbResponse.region === region);
+
+      let logInfo = {
+        workerType: dbResponse.workertype,
+        region: dbResponse.region,
+        az: dbResponse.az,
+        id: id,
+        instanceType: dbResponse.instancetype,
+        srid: dbResponse.srid,
+        imageId: dbResponse.imageId,
+        launched: dbResponse.launched,
+        state,
+        lastevent: generated,
+        metadataSource: 'db',
+      };
+      
+      if (dbResponse.lastevent < generated) {
+        try {
+          if (state === 'pending' || state === 'running') {
```

Comment:
Do we have a race here where an instance could be removed from the table if it's in a "rebooting" state, but then we try to spin up more instances because it looks like we have less running capacity than what we might really have? Some windows worker types reboot after every task, and at least one during the initial setup before claiming work.